File: src/utils/board_utils.py
from typing import List, Set
from ..models.board import Board
from ..models.types import Direction
from typing import List, Set, Tuple, Optional

class BoardUtils:
    """Utilitaires pour manipuler le plateau de jeu."""
    
    @staticmethod
    def get_prefix(board: Board, row: int, col: int, direction: Direction) -> str:
        """Obtient le préfixe pour une position donnée."""
        prefix = []
        current_row, current_col = row, col
        
        while True:
            if direction == Direction.HORIZONTAL:
                current_col -= 1
            else:
                current_row -= 1
                
            if not (0 <= current_row < board.size and 
                   0 <= current_col < board.size):
                break
                
            letter = board.get_letter(current_row, current_col)
            if not letter:
                break
            prefix.insert(0, letter)
            
        return ''.join(prefix)
    
    @staticmethod
    def get_suffix(board: Board, row: int, col: int, direction: Direction) -> str:
        """Obtient le suffixe pour une position donnée."""
        suffix = []
        current_row, current_col = row, col
        
        while True:
            if direction == Direction.HORIZONTAL:
                current_col += 1
            else:
                current_row += 1
                
            if not (0 <= current_row < board.size and 
                   0 <= current_col < board.size):
                break
                
            letter = board.get_letter(current_row, current_col)
            if not letter:
                break
            suffix.append(letter)
            
        return ''.join(suffix)
    
    def check_word_placement(self, board: Board, word: str, row: int, col: int, direction: Direction) -> bool:
        """Vérifie les règles de base pour le placement d'un mot."""
        # 1. Vérifie les limites du plateau
        word_length = len(word)
        if direction == Direction.HORIZONTAL:
            if col < 0 or col + word_length > board.size:
                return False
        else:  # VERTICAL
            if row < 0 or row + word_length > board.size:
                return False

        # 2. Vérifie le premier coup
        if board.is_empty():
            center = board.size // 2
            if direction == Direction.HORIZONTAL:
                return row == center and (col <= center < col + word_length)
            else:  # VERTICAL
                return col == center and (row <= center < row + word_length)

        # 3. Vérifie les connexions
        found_connection = False
        
        for i, letter in enumerate(word):
            current_row = row + (i if direction == Direction.VERTICAL else 0)
            current_col = col + (i if direction == Direction.HORIZONTAL else 0)
            
            existing = board.get_letter(current_row, current_col)
            if existing:
                if existing != letter:
                    return False
                found_connection = True
            elif board.is_adjacent_to_letter(current_row, current_col):
                found_connection = True

        return found_connection or (board.is_empty() and word_length > 0)

File: frontend_new/src/engine/utils/BoardUtils.ts
import { Board } from '../models/Board';
import { Direction } from '../models/Types';

export class BoardUtils {
    /**
     * Obtient le préfixe pour une position donnée.
     */
    static getPrefix(board: Board, row: number, col: number, direction: Direction): string {
        const prefix: string[] = [];
        let currentRow = row;
        let currentCol = col;

        while (true) {
            if (direction === Direction.HORIZONTAL) {
                currentCol -= 1;
            } else {
                currentRow -= 1;
            }

            if (!(0 <= currentRow && currentRow < board.size && 0 <= currentCol && currentCol < board.size)) {
                break;
            }

            const letter = board.getLetter(currentRow, currentCol);
            if (!letter) {
                break;
            }
            prefix.unshift(letter);
        }

        return prefix.join('');
    }

    /**
     * Obtient le suffixe pour une position donnée.
     */
    static getSuffix(board: Board, row: number, col: number, direction: Direction): string {
        const suffix: string[] = [];
        let currentRow = row;
        let currentCol = col;

        while (true) {
            if (direction === Direction.HORIZONTAL) {
                currentCol += 1;
            } else {
                currentRow += 1;
            }

            if (!(0 <= currentRow && currentRow < board.size && 0 <= currentCol && currentCol < board.size)) {
                break;
            }

            const letter = board.getLetter(currentRow, currentCol);
            if (!letter) {
                break;
            }
            suffix.push(letter);
        }

        return suffix.join('');
    }

    /**
     * Vérifie les règles de base pour le placement d'un mot.
     */
    static checkWordPlacement(board: Board, word: string, row: number, col: number, direction: Direction): boolean {
        const wordLength = word.length;

        // 1. Vérifie les limites du plateau
        if (direction === Direction.HORIZONTAL) {
            if (col < 0 || col + wordLength > board.size) {
                return false;
            }
        } else {
            if (row < 0 || row + wordLength > board.size) {
                return false;
            }
        }

        // 2. Vérifie le premier coup
        if (board.isEmpty()) {
            const center = Math.floor(board.size / 2);
            if (direction === Direction.HORIZONTAL) {
                return row === center && (col <= center && center < col + wordLength);
            } else {
                return col === center && (row <= center && center < row + wordLength);
            }
        }

        // 3. Vérifie les connexions
        let foundConnection = false;

        for (let i = 0; i < wordLength; i++) {
            const letter = word[i];
            const currentRow = row + (direction === Direction.VERTICAL ? i : 0);
            const currentCol = col + (direction === Direction.HORIZONTAL ? i : 0);

            const existing = board.getLetter(currentRow, currentCol);
            if (existing) {
                if (existing !== letter) {
                    return false;
                }
                foundConnection = true;
            } else if (board.isAdjacentToLetter(currentRow, currentCol)) {
                foundConnection = true;
            }
        }

        return foundConnection || (board.isEmpty() && wordLength > 0);
    }
}
